drakrun/drakrun/run.py

```diff
@@ -37,39 +37,6 @@ def run_vm(vm_id):
         with open("/var/log/xen/qemu-dm-vm-{vm_id}.log".format(vm_id=vm_id), "rb") as f:
             logging.error(f.read())
 
-    subprocess.run(["xl", "qemu-monitor-command",
-                    "vm-{vm_id}".format(vm_id=vm_id),
-                    "device_del nic0"], check=True)
-
-    for _ in range(120):
-        output = subprocess.check_output(["xl", "qemu-monitor-command",
-                                          "vm-{vm_id}".format(vm_id=vm_id),
-                                          "info network"])
-
-        if b"nic0" not in output:
-            break
-
-        time.sleep(0.25)
-    else:
-        raise RuntimeError("Failed to see nic0 deleted in vm-{vm_id}".format(vm_id=vm_id))
-
-    vm_mac = print_mac(gen_mac(vm_id))
-    subprocess.run(["xl", "qemu-monitor-command",
-                    "vm-{vm_id}".format(vm_id=vm_id),
-                    "device_add e1000,id=nic1,netdev=net0,mac={vm_mac}".format(vm_mac=vm_mac)], check=True)
-
-    for _ in range(120):
-        output = subprocess.check_output(["xl", "qemu-monitor-command",
-                                          "vm-{vm_id}".format(vm_id=vm_id),
-                                          "info network"])
-
-        if b"nic1" in output:
-            break
-
-        time.sleep(0.25)
-    else:
-        raise RuntimeError("Failed to see nic1 created in vm-{vm_id}".format(vm_id=vm_id))
-
```

Comment:
without this hack we would have the same, fixed MAC address every time; maybe this should be re-added

Reply:
TODO check if there is any problem with fixed MAC in current setup (each VM has different interface so probably no?)

Reply:
checked - this doesn't cause any trouble itself, just few more config options are needed to make multiple `dnsmasq` work properly

Reply:
@tklengyel Do you have some experience with network issues when doing a VM cloning? Were you just leaving the same MAC for each VM or doing some other stuff to fix it?

Reply:
Yes. MAC address collision is a problem. At one point I wrote my own NAT engine to route based on vlan id instead of MAC/IP using Honeybrid. Even if you have separate bridges for each clone, each clone will need a unique IP and MAC otherwise regular iptables and arptables get confused. You can inject a command into each clone to reconfigure those.

Reply:
> Even if you have separate bridges for each clone, each clone will need a unique IP and MAC otherwise regular iptables and arptables get confused.

@tklengyel FYI As far as I'm testing right now, arptables are getting it right, at least at Debian Buster:
```
root@debian:/home/mq# arp
Address                  HWtype  HWaddress           Flags Mask            Iface
10.13.12.140             ether   00:16:3e:4f:02:1c   C                     drak12
192.168.121.254          ether   00:50:56:e1:1f:2c   C                     ens33
10.13.19.140             ether   00:16:3e:4f:02:1c   C                     drak19
192.168.121.2            ether   00:50:56:e8:ca:a9   C                     ens33
192.168.121.1            ether   00:50:56:c0:00:08   C                     ens33
```

the Internet seems to be functional. I've tested both running `ping -t 8.8.8.8` and running Internet Explorer simultaneously on both VMs.

I'm not sure if I understand networking at all (:D) but theoretically, this setup doesn't have any MAC collisions within a single hop, so there is a chance for it to work. On the IP level there are no collisions because each VM would get the IP address from different subnet.

So I think right now we'll stick with this solution.

Reply:
Yeap, I guess as long as the ARP is not the same on the same bridge it should work. Since the IPs are on different subnets ip routing will send the packet down the right bridge, then look up the ARP there. How do you reconfigure the IP in the clone?

Reply:
We are just injecting `ipconfig /renew` before the actual malware is executed.

Another trick is to start the guest VM with a bridge that doesn't have a functional DHCP, reset the network card through Windows GUI and make snapshot right afterwards. The VM will ask for an IP address right after being restored. This approach seems to be pretty stable but it requires either:

* instructing a human who makes snapshot (prone to human errors)
* providing some automation that would inject `ipconfig /release && ipconfig /renew` right before snapshotting (this is some interesting direction but using injector on a VM which is about to be snapshotted is pretty risky)

so we are not using this approach right now.